File: data-structures/java/data-structures/src/main/java/com/data_structures/list/linked_list/xor_linked_list/XorLinkedList.java
package com.data_structures.list.linked_list.xor_linked_list;

import java.util.HashMap;

class XorLinkedList<Element> implements IXorLinkedList<Element> {
    private XorLinkedListNode<Element> head;
    private XorLinkedListNode<Element> tail;

    private HashMap<Integer, XorLinkedListNode<Element>> nodes;

    XorLinkedList() {
        head = null;
        tail = null;
        nodes = new HashMap<Integer, XorLinkedListNode<Element>>();
    }

    private int _xor(int a, int b) {
        // Helper function to get the XOR of two IDs
        return a ^ b;
    }

    public void insertAtHead(Element data) {
        XorLinkedListNode newNode = new XorLinkedListNode(data);
        int newId = System.identityHashCode(newNode);
        nodes.put(newId, newNode);
        if (head != null) {
            newNode.both = _xor(0, System.identityHashCode(head));
            head.both = _xor(System.identityHashCode(newNode), 0);
        } else {
            tail = newNode;
        }
        head = newNode;
    }

    public void insertAtTail(int data) {
        // Inserts a new node with the provided data at the tail of the list
        XorLinkedListNode newNode = new XorLinkedListNode(data);
        int newId = System.identityHashCode(newNode);
        nodes.put(newId, newNode);

        if (tail != null) {
            // Adjusting both values for the new node and existing tail
            newNode.both = _xor(System.identityHashCode(tail), 0);
            tail.both = _xor(newNode.both, System.identityHashCode(tail));
        } else {
            // If the list is empty, the new node becomes the head
            head = newNode;
        }
        tail = newNode;
    }

    public void print() {
        XorLinkedListNode<Element> current = this.head;
        int prev = 0;
        while (current != null) {
            System.out.print(current.data + " => ");
            int next = _xor(prev, current.both);
            prev = System.identityHashCode(current);
            current = nodes.getOrDefault(next, null);
        }
    }

    public static void main(String[] args) {
        XorLinkedList<Integer> list = new XorLinkedList<Integer>();
        list.insertAtHead(10);
        list.insertAtHead(20);
        list.insertAtHead(30);
        list.insertAtHead(40);

        list.print(); // Expected: 20 10 30 40
    }
}

File: data-structures/python/data-structures/src/list/linked_list/xor_linked_list/XOR_Linked_List.py
from src.list.linked_list.xor_linked_list.XOR_Linked_List_Node import XORLinkedListNode
from src.list.linked_list.xor_linked_list.IXOR_Linked_List import IXORLinkedList


class XORLinkedList(IXORLinkedList):

    def __init__(self):
        self._head = None
        self._tail = None
        self.id_to_node = {}

    @property
    def head(self):
        return self._head

    @head.setter
    def head(self, val):
        self._head = val

    @head.deleter
    def head(self):
        del self._head

    @property
    def tail(self):
        return self._tail

    @tail.setter
    def tail(self, val):
        self._tail = val

    @tail.deleter
    def tail(self):
        del self._tail

    def XOR(self, a, b):
        return a ^ b

    def add_data(self, data):
        new_node = XORLinkedListNode(data)
        if self.head is None:
            self.head = self.tail = new_node
        else:
            new_node.both = id(self.tail)
            self.tail.both ^= id(new_node)
            self.tail = new_node

    def traverse(self):
        current = self.head
        previous_address = 0

        while current:
            print(current.data, end=" ")
            next_address = self.XOR(previous_address, current.npx)
            previous_address = id(current)
            current = self.get_node_from_address(next_address)

    def get_node_from_address(self, address):
        return None if address == 0 else self.id_to_node.get(address)


xor_linked_list = XORLinkedList()
xor_linked_list.add_data(1)
xor_linked_list.add_data(2)
xor_linked_list.add_data(3)
xor_linked_list.traverse()
